protocols/identify/src/id_transport.rs

```diff
@@ -0,0 +1,237 @@
+        let muxer = Arc::new(muxer);
+        let opening = muxing::outbound_from_ref_and_wrap(muxer.clone());
+
+        IdRetreiver {
+            muxer: Some(muxer),
+            state: IdRetreiverState::OpeningSubstream(opening, config),
+            endpoint,
+        }
+    }
+}
+
+impl<TMuxer> Future for IdRetreiver<TMuxer>
+where TMuxer: muxing::StreamMuxer + Send + Sync + 'static,
+      TMuxer::Substream: Send,
+{
+    type Item = (PeerId, TMuxer);
+    type Error = IoError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // This loop is here so that we can continue polling until we're ready.
+        loop {
+            // In order to satisfy the borrow checker, we extract the state and temporarily put
+            // `Poisoned` instead.
+            match mem::replace(&mut self.state, IdRetreiverState::Poisoned) {
+                IdRetreiverState::OpeningSubstream(mut opening, config) => {
+                    match opening.poll() {
+                        Ok(Async::Ready(Some(substream))) => {
+                            let upgrade = apply::apply(substream, config, self.endpoint);
+                            self.state = IdRetreiverState::NegotiatingIdentify(upgrade);
+                        },
+                        Ok(Async::Ready(None)) => {
+                            return Err(IoError::new(IoErrorKind::Other, "remote refused our identify attempt"));
+                        },
+                        Ok(Async::NotReady) => {
+                            self.state = IdRetreiverState::OpeningSubstream(opening, config);
+                            return Ok(Async::NotReady);
+                        },
+                        Err(err) => return Err(err),
+                    }
+                },
+                IdRetreiverState::NegotiatingIdentify(mut nego) => {
+                    match nego.poll() {
+                        Ok(Async::Ready(IdentifyOutput::RemoteInfo { info, .. })) => {
+                            self.state = IdRetreiverState::Finishing(info.public_key);
+                        },
+                        Ok(Async::Ready(IdentifyOutput::Sender { .. })) => {
+                            unreachable!("IdentifyOutput::Sender can never be the output from \
+                                          the dialing side");
+                        },
+                        Ok(Async::NotReady) => {
+                            self.state = IdRetreiverState::NegotiatingIdentify(nego);
+                            return Ok(Async::NotReady);
+                        },
+                        Err(err) => return Err(err),
+                    }
+                },
+                IdRetreiverState::Finishing(public_key) => {
+                    // Here is a tricky part: we need to get back the muxer in order to return
+                    // it, but it is in an `Arc`.
+                    let muxer = self.muxer.take()
```

Comment:
This could be simplified by putting the muxer into the `IdRetreiverState` enum. There is no need for an `Option` then and consequently no `take` is required.